src/core/server/saved_objects/service/lib/aggregations/validation.ts

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { estypes } from '@elastic/elasticsearch';
+import { ObjectType } from '@kbn/config-schema';
+import { isPlainObject } from 'lodash';
+
+import { IndexMapping } from '../../../mappings';
+import {
+  isObjectTypeAttribute,
+  rewriteObjectTypeAttribute,
+  isRootLevelAttribute,
+  rewriteRootLevelAttribute,
+} from './validation_utils';
+import { aggregationSchemas } from './aggs_types';
+
+const aggregationKeys = ['aggs', 'aggregations'];
+
+interface ValidationContext {
+  allowedTypes: string[];
+  indexMapping: IndexMapping;
+  currentPath: string[];
+}
+
+/**
+ * Validate an aggregation structure against the declared mappings and
+ * aggregation schemas, and rewrite the attribute fields using the KQL-like syntax
+ * - `{type}.attributes.{attribute}` to `{type}.{attribute}`
+ * - `{type}.{rootField}` to `{rootField}`
+ *
+ * throws on the first validation error if any is encountered.
+ */
+export const validateAndConvertAggregations = (
+  allowedTypes: string[],
+  aggs: Record<string, estypes.AggregationContainer>,
+  indexMapping: IndexMapping
+): Record<string, estypes.AggregationContainer> => {
+  return validateAggregations(aggs, {
+    allowedTypes,
+    indexMapping,
+    currentPath: [],
+  });
+};
+
+/**
+ * Validate a record of aggregation containers,
+ * Which can either be the root level aggregations (`SearchRequest.body.aggs`)
+ * Or a nested record of aggregation (`SearchRequest.body.aggs.myAggregation.aggs`)
+ */
+const validateAggregations = (
+  aggregations: Record<string, estypes.AggregationContainer>,
+  context: ValidationContext
+) => {
+  return Object.entries(aggregations).reduce((memo, [aggrName, aggrContainer]) => {
+    return {
+      ...memo,
+      [aggrName]: validateAggregation(aggrContainer, childContext(context, aggrName)),
+    };
+  }, {} as Record<string, estypes.AggregationContainer>);
```

Comment:
Super-NIT: It's way more performant as
```suggestion
  return Object.fromEntries(
    Object.entries(aggregations).map(([aggrName, aggrContainer]) => {
      return [aggrName, validateAggregation(aggrContainer, childContext(context, aggrName))],
    })
  );
```

Reply:
Seems the fastest is using `reduce` while re-using the memo object (https://www.measurethat.net/Benchmarks/Show/9268/0/objectfromentries-vs-reduce), which I can do here